File: hadoop-hdfs-project/hadoop-hdfs-httpfs/src/main/java/org/apache/hadoop/lib/server/Service.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.hadoop.lib.server;

import org.apache.hadoop.classification.InterfaceAudience;

/**
 * Service interface for components to be managed by the {@link Server} class.
 */
@InterfaceAudience.Private
public interface Service {

  /**
   * Initializes the service. This method is called once, when the
   * {@link Server} owning the service is being initialized.
   *
   * @param server the server initializing the service, give access to the
   * server context.
   *
   * @throws ServiceException thrown if the service could not be initialized.
   */
  public void init(Server server) throws ServiceException;

  /**
   * Post initializes the service. This method is called by the
   * {@link Server} after all services of the server have been initialized.
   *
   * @throws ServiceException thrown if the service could not be
   * post-initialized.
   */
  public void postInit() throws ServiceException;

  /**
   * Destroy the services.  This method is called once, when the
   * {@link Server} owning the service is being destroyed.
   */
  public void destroy();

  /**
   * Returns the service dependencies of this service. The service will be
   * instantiated only if all the service dependencies are already initialized.
   *
   * @return the service dependencies.
   */
  public Class[] getServiceDependencies();

  /**
   * Returns the interface implemented by this service. This interface is used
   * the {@link Server} when the {@link Server#get(Class)} method is used to
   * retrieve a service.
   *
   * @return the interface that identifies the service.
   */
  public Class getInterface();

  /**
   * Notification callback when the server changes its status.
   *
   * @param oldStatus old server status.
   * @param newStatus new server status.
   *
   * @throws ServiceException thrown if the service could not process the status change.
   */
  public void serverStatusChange(Server.Status oldStatus, Server.Status newStatus) throws ServiceException;

}

File: hadoop-yarn-project/hadoop-yarn/hadoop-yarn-capacity-scheduler-ui/src/main/webapp/src/features/validation/service.ts
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


import {
  mergeStagedConfig,
  applyFieldPreview,
  buildEffectivePropertyKey,
} from '~/utils/configUtils';
import {
  runFieldValidation,
  type ValidationContext as RuleContext,
} from '~/config/validation-rules';
import type { ValidationIssue, SchedulerInfo, StagedChange } from '~/types';
import { SPECIAL_VALUES } from '~/types/constants/special-values';
import { dedupeIssues } from './utils/dedupeIssues';

type Severity = ValidationIssue['severity'];

export interface FieldValidationOptions {
  queuePath: string;
  fieldName: string;
  value: unknown;
  configData: Map<string, string>;
  stagedChanges: StagedChange[];
  schedulerData?: SchedulerInfo | null;
}

export interface QueueValidationOptions {
  queuePath: string;
  properties: Record<string, string>;
  configData: Map<string, string>;
  stagedChanges: StagedChange[];
  schedulerData?: SchedulerInfo | null;
}

export interface ValidationResult {
  valid: boolean;
  issues: ValidationIssue[];
}

export function validateField(options: FieldValidationOptions): ValidationResult {
  const { queuePath, fieldName, value, configData, stagedChanges, schedulerData } = options;

  const withStaged = mergeStagedConfig(configData, stagedChanges);
  const effectiveConfig = applyFieldPreview(withStaged, queuePath, fieldName, value);

  const issues = runFieldValidation(
    buildRuleContext({
      queuePath,
      fieldName,
      fieldValue: value,
      config: effectiveConfig,
      schedulerData,
      stagedChanges,
    }),
  );

  return {
    valid: issues.every((issue) => issue.severity !== 'error'),
    issues,
  };
}

export function validateQueue(options: QueueValidationOptions): ValidationResult {
  const { queuePath, properties, configData, stagedChanges, schedulerData } = options;

  const withStaged = mergeStagedConfig(configData, stagedChanges);
  const effectiveConfig = new Map(withStaged);

  Object.entries(properties).forEach(([field, value]) => {
    const key = buildEffectivePropertyKey(queuePath, field);
    if (value === '' || value === null || value === undefined) {
      effectiveConfig.delete(key);
    } else {
      effectiveConfig.set(key, value);
    }
  });

  const fieldsToValidate = new Set<string>(Object.keys(properties));
  if (!fieldsToValidate.has('capacity')) {
    fieldsToValidate.add('capacity');
  }

  const issues: ValidationIssue[] = [];

  fieldsToValidate.forEach((field) => {
    const key = buildEffectivePropertyKey(queuePath, field);
    const fieldValue = Object.prototype.hasOwnProperty.call(properties, field)
      ? properties[field]
      : (effectiveConfig.get(key) ?? '');

    const context = buildRuleContext({
      queuePath,
      fieldName: field,
      fieldValue,
      config: effectiveConfig,
      schedulerData,
      stagedChanges,
    });

    issues.push(...runFieldValidation(context));
  });

  return {
    valid: issues.every((issue) => issue.severity !== 'error'),
    issues: dedupeIssues(issues),
  };
}

function buildRuleContext(params: {
  queuePath: string;
  fieldName: string;
  fieldValue: unknown;
  config: Map<string, string>;
  schedulerData?: SchedulerInfo | null;
  stagedChanges: StagedChange[];
}): RuleContext {
  const { queuePath, fieldName, fieldValue, config, schedulerData, stagedChanges } = params;
  const legacyModeEnabled = config.get(SPECIAL_VALUES.LEGACY_MODE_PROPERTY) !== 'false';

  return {
    queuePath,
    fieldName,
    fieldValue,
    config,
    schedulerData,
    stagedChanges,
    legacyModeEnabled,
  };
}

export function hasBlockingIssues(issues: ValidationIssue[]): boolean {
  return issues.some((issue) => issue.severity === 'error');
}

export function splitIssues(issues: ValidationIssue[]): {
  errors: ValidationIssue[];
  warnings: ValidationIssue[];
} {
  const errors: ValidationIssue[] = [];
  const warnings: ValidationIssue[] = [];

  issues.forEach((issue) => {
    if (issue.severity === 'error') {
      errors.push(issue);
    } else {
      warnings.push(issue);
    }
  });

  return { errors, warnings };
}
